src/components/application_manager/src/state_controller_impl.cc

```diff
@@ -245,129 +245,185 @@ void StateControllerImpl::SetRegularState(ApplicationSharedPtr app,
+      IsStreamableHMIType(applied_hmi_level) && applied_->IsVideoApplication();
 
-  const bool applied_is_streamable = applied_grabs_audio || applied_grabs_video;
+  // app to resolve constants
+  const mobile_apis::HMILevel::eType to_resolve_hmi_level =
+      state_to_resolve->hmi_level();
 
-  // for app to resolve constants
   const bool to_resolve_grabs_audio =
-      HasStreamingPermissions(reg_state_to_resolve->hmi_level()) &&
+      IsStreamableHMIType(to_resolve_hmi_level) &&
       app_to_resolve->IsAudioApplication();
 
   const bool to_resolve_grabs_video =
-      HasStreamingPermissions(reg_state_to_resolve->hmi_level()) &&
+      IsStreamableHMIType(to_resolve_hmi_level) &&
       app_to_resolve->IsVideoApplication();
 
-  const bool to_resolve_is_streamable =
-      to_resolve_grabs_audio || to_resolve_grabs_video;
-
-  const bool to_resolve_handles_full =
-      reg_state_to_resolve->hmi_level() == mobile_apis::HMILevel::HMI_FULL;
+  // compatibility constants
+  const bool same_app_audio_type =
+      IsSameAudioAppType(*applied_, *app_to_resolve);
 
-  const bool same_app_type =
-      state_ctrl_->IsSameAppType(applied_, app_to_resolve);
+  // result vars
+  mobile_apis::VideoStreamingState::eType result_video_state =
+      mobile_apis::VideoStreamingState::NOT_STREAMABLE;
+  mobile_apis::AudioStreamingState::eType result_audio_state =
+      mobile_apis::AudioStreamingState::NOT_AUDIBLE;
 
-  mobile_apis::HMILevel::eType result_hmi_level =
-      reg_state_to_resolve->hmi_level();
-
-  if (applied_grabs_full && to_resolve_is_streamable && !same_app_type) {
-    result_hmi_level = mobile_apis::HMILevel::HMI_LIMITED;
+  if (to_resolve_grabs_audio && !same_app_audio_type) {
+    result_audio_state = mobile_apis::AudioStreamingState::AUDIBLE;
   }
 
-  if ((applied_grabs_full && to_resolve_handles_full &&
-       !to_resolve_is_streamable) ||
-      (applied_is_streamable && to_resolve_is_streamable && same_app_type)) {
-    result_hmi_level = mobile_apis::HMILevel::HMI_BACKGROUND;
+  if (to_resolve_grabs_video && !applied_grabs_video) {
+    result_video_state = mobile_apis::VideoStreamingState::STREAMABLE;
   }
 
-  if (reg_state_to_resolve->hmi_level() != result_hmi_level) {
-    mobile_apis::VideoStreamingState::eType result_video_state =
-        mobile_apis::VideoStreamingState::NOT_STREAMABLE;
-    mobile_apis::AudioStreamingState::eType result_audio_state =
-        mobile_apis::AudioStreamingState::NOT_AUDIBLE;
-
-    if (result_hmi_level == mobile_apis::HMILevel::HMI_LIMITED) {
-      if (to_resolve_grabs_video && !applied_grabs_video) {
-        result_video_state = mobile_apis::VideoStreamingState::STREAMABLE;
-      }
+  mobile_apis::HMILevel::eType result_hmi_level = state_to_resolve->hmi_level();
 
-      if (to_resolve_grabs_audio) {
-        result_audio_state = mobile_apis::AudioStreamingState::AUDIBLE;
-      }
+  if (applied_hmi_level == mobile_apis::HMILevel::HMI_FULL) {
```

Comment:
@EKuliiev Swap comparables in if statements, please. Put rvalue before `==`

Reply:
@EKuliiev please answer the comments if it fixed or of you have concerns not to fix it